### DISCUSSION TITLE
LTIModels with preset Gramians and HSVs. Ideas?

### DISCUSSION THREAD

**Discussion by artpelling** [pymor/pymor#1570] [Ideas]
When constructing LTI Models with system identification methods (ERA and QuadBT specifically), the resulting realizations are balanced and the Hankel singular values of the system are already known from the SVD step in these methods.

I would like to create `LTIModels` from system matrices A,B,C and preset the hankel singular values and gramians of the system such that for example `LTIModel.gramian` just returns the preset value and unecessary computation of a Lyapunov equation solution is avoided.

Does anyone have general thoughts on this?

I'm looking for the least invasive way to enable such a behaviour. I don't think this would be very problematic since the models are immutable.
Subclassing `LTIModel` and adding the hsvs and gramians as properties seems overkill, I would prefer to modify the base class but I don't see a way without interfering with the present structure.

**Comment by pmli** [comment]
I was actually thinking about a similar thing, but concerning system norms (e.g., if it is known that the <i>H</i><sub>&infin;</sub> norm is achieved at zero, this could be set in advance).

Probably the simplest approach would be to add private attributes to `LTIModel` (for pyMOR's immutable objects, attributes starting with an underscore are mutable) that `gramian` and `hsv` methods would be aware of. It is not an elegant approach, because it conflicts with the idea of immutability, which could lead to strange behavior, but strange behavior could also be caused by a user supplying incorrect values of Gramians and Hankel singular values.

The more proper way would be to add public attributes to `LTIModel` which would then have to be set through `__init__`. Although, there would still be the issue that `VectorArrays` and NumPy arrays are usually mutable.

Another aspect is how to support parametric `LTIModels`. Gramians and Hankel singular values should then be parametric objects, but I'm not sure it is feasible (I would expect the parametric dependence is quite complicated except in very simple cases). So, maybe only adding support for non-parametric `LTIModels` is sufficient as a first step.

**Reply by artpelling** [reply]
> The more proper way would be to add public attributes to LTIModel which would then have to be set through __init__.

If this avenue is pursuied thoroughly, it makes sense to implement all attributes in this way (poles, Hankel singular values vectors, norms, gramians...). Wouldn't this lead to a lot of computations in `__init__`? I think this approach is only viable if some sort of lazy evaluation scheme is implemented (only execute the computations, if the attribute is queried and it is not cached yet).
I am not too familiar with pyMOR's `core` and `cache` yet. Do you think this can be implemented with the available tools?

**Reply by pmli** [reply]
Right, basically, `cached` methods simply store their input values and corresponding return values. So, we could do something like
```python
def __init__(self, ..., gramian_data=None, ...):
    # self.__auto_init(locals()) will set self.gramian_data
    ...

def gramian(self, ...):
    if self.gramian_data is not None:
        return self.gramian_data
    # normal computation
    ...
```

**Reply by artpelling** [reply]
I have started a draft in #1583 by adding the option to supply preset values for gramians. Not sure if the type checking is most elegant. I went with a `presets` dict argument, because I think a dedicated preset variable for all cached attributes will be too messy. Of course the name should be subject to discussion. What do you think about the broad strokes?
